spot-client/src/common/remote-control/P2PSignalingBase.js

```diff
@@ -0,0 +1,336 @@
+import { logger } from 'common/logger';
+import { PeerConnection } from 'common/webrtc';
+
+
+/**
+ * @typedef {Object} P2PSignalingCallbacks
+ * @property {function} onReadyStateChanged - Callback called with a boolean value which indicates the readiness state
+ * of a P2P signalling channel.
+ * @property {function} onSendP2PMessage - Callback called when a message needs to be transferred over to the remote P2P
+ * signaling instance during p2p connection establishment process(offer, answer and ICE candidates).
+ * @property {function} onRemoteControlMessageReceived - When a remote control message is received over the P2P
+ * signaling channel.
+ */
+/**
+ * @typedef {Object} P2PSignalingOptions
+ * @property {RTCIceServer[]} iceServers - An array of TURN/STUN servers. See
+ * {@link https://developer.mozilla.org/en-US/docs/Web/API/RTCIceServer/urls} for more details about the format.
+ */
+/**
+ * A peer to peer(direct) signaling channel used to send/receive remote control commands.
+ */
+export default class P2PSignalingBase {
+    /**
+     * Creates new instance.
+     *
+     * @param {P2PSignalingCallbacks} callbacks - The callbacks structure, see type description for more info.
+     * @param {P2PSignalingOptions} options - Extra options, see type description for more info.
+     */
+    constructor(callbacks, options) {
+        this._callbacks = callbacks;
+        this._options = options;
+
+        /**
+         * Stores active {@link PeerConnection}s mapped by remote address string.
+         *
+         * @type {Map<string, PeerConnection>}
+         * @private
+         */
+        this._peerConnections = new Map();
+    }
+
+    /**
+     * Find connection for remote address.
+     *
+     * @param {string} remoteAddress - The remote address string.
+     * @returns {?PeerConnection}
+     */
+    getConnectionForAddress(remoteAddress) {
+        return this._peerConnections.get(remoteAddress);
+    }
+
+    /**
+     * A convenience selector to find the first {@link PeerConnection}.
+     *
+     * @returns {?PeerConnection}
+     */
+    getFirstPeerConnection() {
+        return this._peerConnections.values().next().value;
+    }
+
+    /**
+     * Creates and initializes new {@link PeerConnection} instance for given remote address.
+     *
+     * @param {string} remoteAddress - The remote address string.
+     * @private
+     * @returns {PeerConnection}
+     * @throws an error if {@link PeerConnection} already exists for the given address.
+     */
+    _initPeerConnection(remoteAddress) {
+        if (this._peerConnections.get(remoteAddress)) {
+            throw new Error(`P2P signaling connection already exists for: ${remoteAddress}`);
```

Comment:
I wonder if it should error. Is there a need to make it known this redundant create call happened?

Reply:
This is a case that is supposed to "never happen". I can log an error and early return from the method instead, but weird things could happen.